capsules/src/nfc.rs

```diff
@@ -0,0 +1,270 @@
+                    rb
+                });
+            });
+        });
+    }
+
+    fn frame_transmitted(&self, buffer: &'static mut [u8], result: ReturnCode) {
+        self.tx_buffer.replace(buffer);
+        self.tx_in_progress.take().map(|appid| {
+            let _ = self.application.enter(appid, |app, _| {
+                app.tx_callback
+                    .map(|mut cb| cb.schedule(result.into(), 0, 0));
+            });
+        });
+    }
+}
+
+impl Driver for NfcDriver<'_> {
+    /// Setup shared buffers.
+    ///
+    /// ### `allow_num`
+    ///
+    /// - `1`: Readable buffer for transmission buffer, if
+    ///        provided buffer length is more than MAX_LENGTH then
+    ///        return EINVAL
+    /// - `2`: Writeable buffer for reception buffer, if
+    ///        provided buffer length is not MAX_LENGTH then
+    ///        return EINVAL
+    fn allow(
+        &self,
+        appid: AppId,
+        allow_num: usize,
+        slice: Option<AppSlice<Shared, u8>>,
+    ) -> ReturnCode {
+        match allow_num {
+            1 => self
+                .application
+                .enter(appid, |app, _| {
+                    if let Some(buf) = &slice {
+                        if buf.len() > MAX_LENGTH {
+                            return ReturnCode::EINVAL;
+                        }
+                    }
+                    app.tx_buffer = slice;
+                    ReturnCode::SUCCESS
+                })
+                .unwrap_or_else(|err| err.into()),
+            2 => self
+                .application
+                .enter(appid, |app, _| {
+                    if let Some(buf) = &slice {
+                        if buf.len() != MAX_LENGTH {
+                            return ReturnCode::EINVAL;
+                        }
+                    }
+                    app.rx_buffer = slice;
+                    ReturnCode::SUCCESS
+                })
+                .unwrap_or_else(|err| err.into()),
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    /// Setup callbacks.
+    ///
+    /// ### `subscribe_num`
+    ///
+    /// - `1`: Frame transmission completed callback
+    /// - `2`: Frame reception completed callback
+    fn subscribe(
+        &self,
+        subscribe_num: usize,
+        callback: Option<Callback>,
+        appid: AppId,
+    ) -> ReturnCode {
+        match subscribe_num {
+            1 => self
+                .application
+                .enter(appid, |app, _| {
+                    app.tx_callback = callback;
+                    ReturnCode::SUCCESS
+                })
+                .unwrap_or_else(|err| err.into()),
+            2 => self
+                .application
+                .enter(appid, |app, _| {
+                    app.rx_callback = callback;
+                    ReturnCode::SUCCESS
+                })
+                .unwrap_or_else(|err| err.into()),
+            _ => ReturnCode::ENOSUPPORT,
+        }
```

Comment:
Why have seperate receive/send callbacks? This makes the user space implementation in Rust tricky as you can't share data between the two.

Reply:
I'm a bit confused as to what exactly do you mean by share data between the two? Do you mean working with one buffer for both operations?

Reply:
From my experience you usually have some struct in userspace that you use for parsing the data you send/receive. My having a separate send and receive callback you can't share that one struct between both callbacks in libtock-rs.